### PR TITLE
Use fortranformat

### PR DESCRIPTION
**Pull request by LiamPattinson**
Instead of using utilities such as `ChunkOutput`, `write_1d`, `write_2d`, `f2s`, etc, uses [fortranformat](https://pypi.org/project/fortranformat/) to handle Fortran formatted IO.

- Reimplements `geqdsk.read` and `geqdsk.write`
- Specifies the G-EQDSK standard more rigorously
- Adds utility functions that can read/write ND Fortran arrays to file

Builds upon PR https://github.com/freegs-plasma/FreeQDSK/pull/3, so may need to be revised if that PR needs to be changed significantly. Makes progress towards resolving Issue https://github.com/freegs-plasma/FreeQDSK/issues/2. Similar to an [open PR to FreeGS](https://github.com/freegs-plasma/freegs/pull/80), although it uses `fortranformat` in a slightly different way (no generators, just uses a consistent file handle directly).

Haven't yet updated A-EQDSK, but it shouldn't take too long to push similar changes there.

### CODE REVIEW

**Comment by LiamPattinson** [comment]
This update has now pretty comprehensively rewritten the library to use fortranformat and more rigorously define the file formats, and I'm planning on making a few more additions later today. I've also expanded the documentation beyond the previous PR. Should this PR take precendence?

Things I still need to finish up:
- Need testing data. Currently testing by writing to a string buffer, then reading info back from it. If both the read and write functions make the same mistake, there's no way for us to know. Any idea where I could get some real G-EQDSK/A-EQDSK files that can be freely used?
- Can add some further fields to the end of A-EQDSK
- AFAIK, the only current users are FreeGS and Pyrokinetics. FreeGS is using some of our private methods and in their `pyproject.toml` ask for  `freeqdsk >= 0.1`, so I can't remove unused utility functions without breaking FreeGS. It'd be nice if those utilities could be ported back over to FreeGS.

Note: I've tested using this build in place of 0.1 with Pyrokinetics and it passed all tests, so that's encouraging!

**Comment by bendudson** [comment]
Thanks @LiamPattinson ! I know there was some discussion among the OMFIT folks about EQDSK formats and FreeGS. I'll try pinging them to see if they can help test this.

**Comment by LiamPattinson** [comment]
- Added some G-EQDSK files in use by Pyrokinetics. Let me know if there's a licensing issue here and I'll revert.
- Added an A-EQDSK file provided by Koki Imada (don't know his GitHub handle!).
- Expanded on the G-EQDSK tests, though still need to add some failure cases.

@bendudson Thanks! I've been having trouble finding detailed standards for A-EQDSK files, and have mostly pieced things together from the FreeGS implementation, [eqtools](https://github.com/PSFCPlasmaTools/eqtools), and the old [BOUT implementation](https://github.com/bendudson/BOUT/blob/master/tools/tokamak_grids/pyGridGen/aeqdsk.py). I'm hoping to add p files soon, and information there is even more scarce. Getting more eyes on the problem will definitely be helpful, especially if it's from people who use the files regularly and have access to the Fortran that generates them.

**Comment by bendudson** [comment]
My experience is that there is quite a lot of variation in real GEQDSK files in the wild. It would be good to have a writer that strictly followed the specification (whatever that is), and a reader that is quite forgiving and can cope with variations.

**Comment by bclyons12** [comment]
The rigorous standard for a gEQDSK is: 
 - 'QPSI' positive
 - 'PSIRZ' increases from axis to edge
 - sign of 'Ip' and 'Bt' are given with respect to a counter-clockwise toroidal angle

One way to think about this is that, depending on the sign of 'Ip' and 'Bt', this corresponds to four possible COCOS representation (1,3,5, and 7).
```
        All use sigma_RpZ=+1 (phi is counterclockwise) and exp_Bp=0 (psi is flux/2.*pi)
        We want
        sign(psi_edge-psi_axis) = sign(Ip)*sigma_Bp > 0  (psi always increases in gEQDSK)
        sign(q) = sign(Ip)*sign(Bt)*sigma_rhotp > 0      (q always positive in gEQDSK)
        ::
            ============================================
            Bt    Ip    sigma_Bp    sigma_rhotp    COCOS
            ============================================
            +1    +1       +1           +1           1
            +1    -1       -1           -1           3
            -1    +1       +1           -1           5
            -1    -1       -1           +1           7
```

**Comment by LiamPattinson** [comment]
I've come across a few variations in units and coordinate systems in G-EQDSK files, but so far all of them I've seen follow the same formatting rules. Am I good to assume the data will be written using the `(5e16.9)` Fortran format, or will the `read` function need to fall back on a more generic regex?

**Comment by bendudson** [comment]
@LiamPattinson I'm not certain, but I think it should be safe to assume that standard Fortran format.

**Comment by LiamPattinson** [comment]
I've finished adding tests for this update. A-EQDSK now recognises some extra fields, but I've seen indications that the last few lines aren't properly standardised and they might mean different things depending on the source of the file.

@ZedThree Please could you take a look over this when you next have time?

File: pyproject.toml [hunk 1]
```diff
@@ -23,22 +24,29 @@ classifiers = [
...
 [project.optional-dependencies]
 tests = [
     "pytest >= 3.3.0",
     "pytest-cov",
 ]
+docs = [
+    "sphinx == 3.5.4",
+    "myst_parser == 0.14.0",
+    "jinja2 < 3.1.0",
+    "sphinx_rtd_theme",
```
Comment (ZedThree) [inline pyproject.toml]: I have a feeling these are just copy-pasted from pyrokinetics :)

The sphinx version is very old and doesn't work with python >= 3.10, and I think myst isn't needed for this project. These two should be changed at least.

This is personal preference, but I think `sphinx-book-theme` looks nicer than the RTD theme.

A couple of suggestions for other useful plugins:

```suggestion
     "sphinx >= 5.3",
     "sphinx-book-theme >= 0.4.0rc1",
     "sphinx-argparse-cli >= 1.10.0",
     "sphinx_autodoc_typehints >= 1.19",
```

File: tests/utils.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+from pathlib import Path
+
+import numpy as np
+
+from freeqdsk import geqdsk
+
+
+def scramble_geqdsk(filename: Path, output: Path, magnitude: float = 0.1) -> None:
+    """
+    Used to redact G-EQDSK data, converting actual data into random numbers with the
```
Comment (ZedThree) [inline tests/utils.py:10]: Neat!

File: freeqdsk/_fileutils.py [hunk 3]
```diff
@@ -127,3 +190,174 @@ def next_value(fh):
...
+                            "Additional elements were detected beyond the end of the "
+                            "requested array. These have been discarded."
+                        )
+                        result = result[: shape[0]]
+            return np.array(result)
+        else:
+            # Read ND arrays as flattened 1D arrays, then reshape to requested shape
+            return read_array((np.prod(shape),), fh, fmt).reshape(shape, order="F")
+    else:
+        raise ValueError("'shape' should be a scalar or a 1D array")
```
Comment (ZedThree) [inline freeqdsk/_fileutils.py]: Minor, but moving this block up and guarding on something like `if shape.ndim not in [0, 1]` means you could drop one level of indentation from the `shape.ndim == 1` block

File: freeqdsk/aeqdsk.py [hunk 4]
```diff
@@ -1,280 +1,582 @@
...
+    """
+    if field.name in data:
+        # Check length of array, if present
+        if field.length_of in data and data[field.name] != len(data[field.length_of]):
+            raise ValueError(
+                f"'{field.name}' should be the length of the array '{field.length_of}'"
+            )
+        return int(data[field.name])
+    else:
+        return len(data.get(field.length_of, []))
```
Comment (ZedThree) [inline freeqdsk/aeqdsk.py]: Move the guard clause to the top:

```suggestion
    if field.name not in data:
        return len(data.get(field.length_of, []))

    # Check length of array, if present
    if field.length_of in data and data[field.name] != len(data[field.length_of]):
        raise ValueError(
            f"'{field.name}' should be the length of the array '{field.length_of}'"
        )
    return int(data[field.name])
```

File: freeqdsk/aeqdsk.py [hunk 5]
```diff
@@ -1,280 +1,582 @@
...
+    if field.name in data:
+        result = np.asarray(data[field.name])
+        if result.ndim == 0:
+            result = np.array([result])
+        # Check length of array, if present
+        if field.has_length in data and len(data[field.name]) != data[field.has_length]:
+            raise ValueError(f"'{field.name}' should have length '{field.has_length}'")
+        return result
+    else:
+        return np.zeros(data.get(field.has_length, 0))
```
Comment (ZedThree) [inline freeqdsk/aeqdsk.py]: Move guard clause to top

File: freeqdsk/aeqdsk.py [hunk 6]
```diff
@@ -294,26 +596,94 @@ def write(data, fh):
...
+        _extended_arrays_1,
+        _extended_arrays_2,
+        _extended_general,
+    ]
+    for field in itertools.chain.from_iterable(extended_blocks):
+        if field.name in data:
+            extended = True
+            break
+
+    if extended:
```
Comment (ZedThree) [inline freeqdsk/aeqdsk.py]: Could bail early and unindent next block?

```suggestion
    if not extended:
        return
```

File: freeqdsk/aeqdsk.py [hunk 7]
```diff
@@ -294,26 +596,94 @@ def write(data, fh):
...
-                # Replace the default function with the value, which may depend on
-                # previously read data
-                default = default(data)
 
-            if key is None:
-                output.newline()  # Ensure on a new line
-            else:
-                output.write(data.get(key, default))
+    # Output first block of general data
+    write_array([_float_field(field, data) for field in _general_block_1], fh, data_fmt)
```
Comment (ZedThree) [inline freeqdsk/aeqdsk.py]: Is it worth wrapping this pattern up into a local helper function? Or are there too many variations?

File: freeqdsk/aeqdsk.py [hunk 8]
```diff
@@ -335,27 +706,268 @@ def read(fh):
...
+        values = read_array(data[field.has_length], fh, data_fmt)
+        data[field.name] = values
+
+    # Read in another general data block
+    extended_values = read_array("all", fh, data_fmt)
+    if len(extended_values) > len(_extended_general):
+        warnings.warn(
+            "Encountered variables at the end of an A-EQDSK file that are not "
+            "recognised by FreeQDSK. Please consider contributing to the project "
+            "and letting us know what these are!"
```
Comment (ZedThree) [inline freeqdsk/aeqdsk.py]: Nice! Worth putting in the link to the repo here?

File: freeqdsk/aeqdsk.py [hunk 9]
```diff
@@ -335,27 +706,268 @@ def read(fh):
...
+    # Zip only includes elements up to the shortest iterable provided. data will not
+    # include elments that FreeQDSK doesn't recognise, nor will it store default values
+    # for elements not in the file.
+    for field, value in zip(_extended_general, extended_values):
+        data[field.name] = value
 
     return data
+
+
+# Dynamic docstring generation
```
Comment (ZedThree) [inline freeqdsk/aeqdsk.py:765]: This is very fancy! It's probably worth putting a short note as to _why_ the docstring is dynamically generated

**Review by ZedThree — APPROVED** [review]
LGTM, a few little comments. I would definitely bump sphinx version at the very least

But very nice, comprehensive tests and docs, thanks!

**Comment by LiamPattinson** [comment]
Thanks for the tips! I've implemented the things you suggested:

- Docs updated. autodoc typehints will make things easier going forwards. It seems it isn't handling `np.typing.ArrayLike` well, but it's only a minor issue.
- Simplified `read_array`. Rewrote it so that a scalar `shape` is the 'standard' path through the function, and all ND arrays go through the same process of reading a flattened array and reshaping.
- Simplified A-EQDSK. Now writes the last line instead of going up to the last field.

**Review by ZedThree — APPROVED** [review]

**Pull request merged**